File: 248_count-of-smaller-number/count-of-smaller-number.cpp
/*
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/count-of-smaller-number
@Language: C++
@Datetime: 16-07-06 10:36
*/

class SegmentTree {
public:
    SegmentTree * left;
    SegmentTree * right;
    int start;
    int end;
    int count;
    
    SegmentTree(int _start, int _end) {
        start = _start; end = _end;
        left = right = NULL;
        count = 0;
    }
    
    static SegmentTree * buildTree(int start, int end) {
        if (start > end) return NULL;
        SegmentTree * root = new SegmentTree(start, end);
        if (start != end) {
            int mid = ((end - start) >> 1) + start;
            root->left = SegmentTree::buildTree(start, mid);
            root->right = SegmentTree::buildTree(mid + 1, end);
        }
        return root;
    }
    
    void inc(int a) {
        if (start == a && end == a) {
            count += 1;
            return ;
        }
        int mid = ((end - start) >> 1) + start;
        if (mid >= a)
            left->inc(a);
        else
            right->inc(a);
        count = left->count + right->count;
    }
    
    int sum(int _start, int _end) {
        if (start > _end || end < _start) {
            return 0;
        }
        if (start >= _start && end <= _end) {
            return count;
        }
        return left->sum(_start, _end) + right->sum(_start, _end);
    }
    
    ~SegmentTree() {
        if (left) 
            delete left;
        if (right) 
            delete right;
    }
};


class Solution {
public:
   /**
     * @param A: An integer array
     * @return: The number of element in the array that 
     *          are smaller that the given integer
     */
    vector<int> countOfSmallerNumber(vector<int> &A, vector<int> &queries) {
        // write your code here
        int m = A.size();
        int n = queries.size();
        vector<int> ret(n, 0);
        if (m == 0)
            return ret;
        int mm = INT_MIN;
        for (auto n : A) {
            mm = max(mm, n);
        }
        
        SegmentTree * root = SegmentTree::buildTree(0, mm);
        for (auto n : A) {
            root->inc(n);
        }
        for (int i = 0; i < n; i++ ) { 
            ret[i] = root->sum(0, queries[i] - 1);
        }
        return ret;
    }
};

File: 248_count-of-smaller-number/count-of-smaller-number.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/count-of-smaller-number
@Language: Python
@Datetime: 16-06-27 15:51
'''

class Solution:
    """
    @param A: A list of integer
    @return: The number of element in the array that 
             are smaller that the given integer
    """
    def countOfSmallerNumber(self, A, queries):
        # write your code here
        A = sorted(A)
        ret = []
        def helper(A, q):
            if not A:
                return 0
            l, r = 0, len(A) - 1
            while l < r:
                mid = (r + l) / 2
                if A[mid] == q:
                    l = mid
                    break
                elif A[mid] > q:
                    r = mid - 1
                else:
                    l = mid + 1 
            while l > 0 and A[l] == q:
                l -= 1
            if A[l] < q:
                return l + 1
            elif A[l] > q:
                return l
            else:
                return 0
            
        for q in queries:
            ret.append(helper(A, q))
        return ret